#7280: Add support for binary division

VirdhatchaniKN opened the pull request:
Add support for Binary Division ( TTNN migration included)
- All post-commit tests : [Link](https://github.com/tenstorrent-metal/tt-metal/actions/runs/8631841004) - PASSED
- Fast Dispatch unit tests : [Link](https://github.com/tenstorrent-metal/tt-metal/actions/runs/8631843008) - PASSED
- Slow Dispatch unit tests : [Link](https://github.com/tenstorrent-metal/tt-metal/actions/runs/8631845355) - PASSED

Test files : [Link](https://github.com/tenstorrent-metal/tt-metal/pull/7287#issuecomment-2056872375) 

TTNN changes removed : [Link](https://github.com/tenstorrent/tt-metal/pull/7287#issuecomment-2063857694)

VirdhatchaniKN committed:
#7280: Add support for binary division

tt-aho commented on tt_eager/tt_dnn/op_library/composite/composite_ops.cpp:
```diff
@@ -834,6 +834,26 @@ Tensor addcdiv(
+    const Tensor& input_a,
+    const Tensor& input_b,
+    const MemoryConfig& output_mem_config) {
+    Tensor result = mul(input_a, recip(input_b, output_mem_config), std::nullopt, output_mem_config);
+    Tensor t_inf = full_like(input_a, std::numeric_limits<float>::infinity(), output_mem_config);
+    Tensor t_nan = full_like(input_a, std::nanf(""), output_mem_config);
+    return where(eqz(input_b, output_mem_config),
+                where(eqz(input_a, output_mem_config),
+                    t_nan,
+                    mul(t_inf, sign(input_a, output_mem_config), std::nullopt, output_mem_config), output_mem_config),
+                result, output_mem_config);
```
I would prefer we at least expose just the mul with recip as its own div op (our eltwise binary implementation supports fusing unary ops to prescale binary inputs). This support for handling nan adds a lot of extra ops and overhead, whereas if the user knows / doesn't care about nans, then having this exposed separately would be much better for perf since it would only be a single op. Not sure what your opinion on this is @eyonland ? This might be true for other ops as well that we could potentially expose a path like this.

We should also review this code for handling nans that is being added everywhere to make it more performant, as well as commonize the logic instead of copy pasting similar code everywhere

VirdhatchaniKN replied:
Hi @tt-aho , @eyonland 
I tried implementing using Unary chain method
```
     case BinaryOpType::BINARY_DIV:
            defines.merge(eltwise_unary_op_utils::get_defines(UnaryOpType::RECIP, std::nullopt, "PRE_IN1_0"));
            op_name = "mul_tiles";
            op_code = "2";
            break;
```

On using this method, the test are failing (assertion error). This is due to hardware restrictions related to nan  as  discussed in issue #4409 . Below shared is the result comparison for Binary_div on implementing in composite structure and unary chain method

- Composite structure ( Current implementation ) - Tests PASSED
<img width="1119" alt="Screenshot 2024-04-11 at 3 44 25 PM" src="https://github.com/tenstorrent-metal/tt-metal/assets/138196495/d26c2101-9ded-46d6-9cc9-36f247f81ffb">

- Unary chain method - Tests FAILED
<img width="1164" alt="Screenshot 2024-04-11 at 3 47 01 PM" src="https://github.com/tenstorrent-metal/tt-metal/assets/138196495/9035409e-cdbb-4e83-b3ee-b40c47ec089b">

Would like your inputs on this.
Thanks.

tt-aho replied:
@VirdhatchaniKN my comment did not say to replace the composite method as I understand that the non-composite version would not produce correct output when denominator is 0. It meant that if user knows/can guarantee the denominator is not 0, then they can avoid a lot of overhead since the current logic to handle nans adds >= 6 additional ops plus overhead of creating the 2 full tensors.

tt-aho commented on tt_eager/tt_dnn/op_library/composite/composite_ops.hpp:
```diff
@@ -160,6 +160,11 @@ Tensor addcdiv(
     float value,
     const MemoryConfig& output_mem_config = operation::DEFAULT_OUTPUT_MEMORY_CONFIG);
 
+Tensor binary_div(
```
Is there a reason we are calling this op `binary_div`?
Our other eltwise binary ops (add, sub, mul) do not have `binary_` in the name?

VirdhatchaniKN replied:
Since both the inputs are tensors , I've named it as binary_div, and moreover there is a seperate op `Tensor div_unary( const Tensor& input_tensor, float value, const MemoryConfig& output_mem_config = operation::DEFAULT_OUTPUT_MEMORY_CONFIG);` . Hence the naming @tt-aho  .

eyonland reviewed (changes requested):
The ttnn api is intended to match the torch api.  We cannot deviate from this.

VirdhatchaniKN commented:
> The ttnn api is intended to match the torch api. We cannot deviate from this.

Made the necessary changes
Test results : 
File : `tests/ttnn/sweep_tests/sweeps/div.py`
<img width="1058" alt="Screenshot 2024-04-15 at 6 35 15 PM" src="https://github.com/tenstorrent-metal/tt-metal/assets/138196495/c63a8313-6097-43cf-b705-b48259c8f60f">

File : `tests/ttnn/unit_tests/operations/test_math_binary.py`
<img width="637" alt="Screenshot 2024-04-15 at 6 36 06 PM" src="https://github.com/tenstorrent-metal/tt-metal/assets/138196495/61ab9be5-0776-4ffb-8b3c-417ed715069f">

File : `tests/tt_eager/python_api_testing/sweep_tests/pytests/tt_dnn/test_composite.py`
<img width="879" alt="Screenshot 2024-04-15 at 6 39 17 PM" src="https://github.com/tenstorrent-metal/tt-metal/assets/138196495/b8cbad43-259a-4789-ae1d-bddb0c1f8441">

Documentation:
<img width="1007" alt="Screenshot 2024-04-15 at 6 45 15 PM" src="https://github.com/tenstorrent-metal/tt-metal/assets/138196495/36f38330-5a1a-4c60-9aa2-5734c602e555">

arakhmati reviewed (changes requested):
Please move `ttnn.div` from ttnn/operations/math.py to ttnn/operations/binary.py

VirdhatchaniKN commented:
> Please move `ttnn.div` from ttnn/operations/math.py to ttnn/operations/binary.py

Made the changes @arakhmati 
<img width="637" alt="Screenshot 2024-04-17 at 2 53 45 PM" src="https://github.com/tenstorrent/tt-metal/assets/138196495/831c21ab-fb74-49d9-bb86-fc4babcb524f">

VirdhatchaniKN commented:
TTNN Changes removed

- Renamed `binary_div` to `div`
- Added `div_fast` to divide non-zero tensors
- Test file : `tests/tt_eager/python_api_testing/sweep_tests/pytests/tt_dnn/test_div.py`
<img width="863" alt="Screenshot 2024-04-18 at 6 35 29 PM" src="https://github.com/tenstorrent/tt-metal/assets/138196495/77dc0ddb-d643-47a3-b5f1-716e6d9fe97c">

VirdhatchaniKN commented on tt_eager/tt_lib/csrc/tt_lib_bindings_tensor_composite_ops.cpp:
```diff
@@ -791,7 +791,23 @@ namespace tt::tt_metal::detail{
+        m_tensor.def("div", &div,
+            py::arg("input_a").noconvert(), py::arg("input_b").noconvert(), py::arg("accurate_mode") = false, py::arg("output_mem_config").noconvert() = operation::DEFAULT_OUTPUT_MEMORY_CONFIG, R"doc(
+            Performs the element-wise division of ``input_a`` by ``input_b``.
+            If the input tensors are non-zero tensors, then ``accurate_mode`` can be set to ``true`` .
```
Yes @tt-aho . Only `input_b` needs to be a non-zero tensor. Else accurate_mode can be set to true.
I’ll update.

eyonland commented on tests/tt_eager/python_api_testing/sweep_tests/pytests/tt_dnn/test_div.py:
```diff
@@ -0,0 +1,79 @@
+
+        run_single_pytorch_test(
+            "eltwise-div",
+            input_shapes,
```
This is not the test for "div".

eyonland commented on tt_eager/tt_lib/csrc/tt_lib_bindings_tensor_composite_ops.cpp:
```diff
@@ -791,7 +791,23 @@ namespace tt::tt_metal::detail{
         )doc");
 
+        m_tensor.def("div", &div,
+            py::arg("input_a").noconvert(), py::arg("input_b").noconvert(), py::arg("accurate_mode") = false, py::arg("output_mem_config").noconvert() = operation::DEFAULT_OUTPUT_MEMORY_CONFIG, R"doc(
```
Where is the test for "div" ?

eyonland commented on tests/tt_eager/python_api_testing/sweep_tests/tt_lib_ops.py:
```diff
@@ -1025,6 +1025,26 @@ def eltwise_addalpha(
+    t0 = setup_tt_tensor(x, device, layout[0], input_mem_config[0], dtype[0])
+    t1 = setup_tt_tensor(y, device, layout[1], input_mem_config[1], dtype[1])
+    t2 = ttl.tensor.div(t0, t1, accurate_mode, output_mem_config=output_mem_config)
+
```
Ok, here it is.

VirdhatchaniKN commented:
- All post-commit tests - [Link](https://github.com/tenstorrent/tt-metal/actions/runs/8832179670) - PASSED
- [post-commit] Fast Dispatch unit tests - [Link](https://github.com/tenstorrent/tt-metal/actions/runs/8832181261) - PASSED
- [post-commit] Slow Dispatch unit tests - [Link](https://github.com/tenstorrent/tt-metal/actions/runs/8832187291) - PASSED
